src/main/java/jenkins/plugins/logstash/configuration/LogstashIndexer.java

```diff
@@ -0,0 +1,172 @@
+package jenkins.plugins.logstash.configuration;
+
+import javax.annotation.Nonnull;
+
+import org.apache.commons.lang.StringUtils;
+import org.kohsuke.stapler.DataBoundSetter;
+import org.kohsuke.stapler.QueryParameter;
+import org.kohsuke.stapler.StaplerRequest;
+
+import hudson.DescriptorExtensionList;
+import hudson.ExtensionPoint;
+import hudson.model.AbstractDescribableImpl;
+import hudson.model.Descriptor;
+import hudson.model.ReconfigurableDescribable;
+import hudson.model.Descriptor.FormException;
+import hudson.util.FormValidation;
+import jenkins.model.Jenkins;
+import jenkins.plugins.logstash.Messages;
+import jenkins.plugins.logstash.persistence.AbstractLogstashIndexerDao;
+import net.sf.json.JSONObject;
+
+/**
+ * Extension point for logstash indexers.
+ * This extension point provides the configuration for the indexer. You also have to implement the actual
+ * indexer in a separate class extending {@link AbstractLogstashIndexerDao}.
+ *
+ * @param <T> The class implementing the push to the indexer
+ */
+public abstract class LogstashIndexer<T extends AbstractLogstashIndexerDao>
+    extends AbstractDescribableImpl<LogstashIndexer<?>>
+    implements ExtensionPoint, ReconfigurableDescribable<LogstashIndexer<?>>
+{
+  protected String host;
+  protected int port;
+
+  protected transient T instance;
+
+  /**
+   * Returns the host for connecting to the indexer.
+   *
+   * @return Host of the indexer
+   */
+  public String getHost()
+  {
+    return host;
+  }
+
+  /**
+   * Sets the host for connecting to the indexer.
+   *
+   * @param host
+   *          host to connect to.
+   */
+  @DataBoundSetter
+  public void setHost(String host)
+  {
+    this.host = host;
+  }
+
+  /**
+   * Returns the port for connecting to the indexer.
+   *
+   * @return Port of the indexer
+   */
+  public int getPort()
+  {
+    return port;
+  }
+
+  /**
+   * Sets the port used for connecting to the indexer
+   *
+   * @param port
+   *          The port of the indexer
+   */
+  @DataBoundSetter
+  public void setPort(int port)
+  {
+    this.port = port;
+  }
+
+  /**
+   * Gets the instance of the actual {@link AbstractLogstashIndexerDao} that is represented by this
+   * configuration.
+   * Checks via {@link #shouldRefreshInstance()}, if a new {@link AbstractLogstashIndexerDao}
+   * needs to be created. If yes calls {@link #createIndexerInstance()} to create a new instance.
+   *
+   * @return {@link AbstractLogstashIndexerDao} instance
+   */
+  @Nonnull
+  public final T getInstance()
```

Comment:
Do we need to synchronize it? I can't find anything via a quick google

Reply:
good question. Potentially after a reconfiguration or initially we might have 2 builds running into the if at the same time. Probably it will not hurt to synchronize. The worst that happens is that 2 builds have 2 different daos with the same configuration. Will this be a problem? We will already have different daos active when configuration is changed (the builds that were running when configuration changed and builds started after change).

Reply:
Not at the moment, but I think it would be nice to enable a `close()` method on the Daos later to let them use long living connections etc.

Reply:
Another problem is somebody calling `setHost()` concurrently with `getInstance()` -- those are public methods after all.

Maybe it's covered by creating a new instance 
> when you bind a request to the configuration you always get a new instance of the logstashindexer

but then I don't actually understand how can caching the Dao instance in the LogstashIndexer work?

Reply:
it is cached because I use the reconfigure instead of the default behaviour of staper to create new instance. But this question is not relevant when we switch to the change in my configuration2 branch.